doc/Yang Model Guidelines.md

```diff
@@ -0,0 +1,248 @@
+#   SONiC YANG MODEL GUIDELINES
+
+### Rev 1.0 ###
+
+### Revision ###
+
+ | Rev |     Date    |       Author       | Change Description                |
+ |:---:|:-----------:|:------------------:|-----------------------------------|
+ | 1.0 | 22 Aug 2019             |      Praveen Chaudhary      | Initial version                    |
+ | 1.1 | 29 Aug 2019             |      Praveen Chaudhary      | Addressed review comments.
+
+
+
+###  This document lists the guidelines, which will be used to write YANG Modules for SONiC. These YANG Modules will be primarily based on or represent the ABNF Schema of SONiC, and the syntax of YANG models must follow RFC 7950.
+
+### These YANG models will be used to verify the configuration for SONiC switches, so a library which supports validation of configuration on SONiC Switch must use these YANG Models. List of such Libraries are: 1.) Configuration Validation Library. (CVL)
+
+
+
+### 1. Each primary section of Config DB (i.e a dictionary in config_db.json) for Example, VLAN, VLAN_MEMBER, INTERFACE  in config DB will be mapped to a container in YANG model.
+
+Example:
+
+#### ABNF
+
+```
+"INTERFACE": {
+        ~~"Ethernet112|2a04:fxxx:40:a709::xxx/126": {
+            "scope": "global",
+            "family": "IPv6"~~
+        }
+ }
+```
+will translate to:
+
+####  YANG
+--
+```
+container INTERFACE {
+~~
+~~
+}
+```
+
+### 2. Each leaf in YANG module should have same name as corresponding key in Config DB.
+
+Example:
+Leaf names are same PACKET_ACTION, IP_TYPE and PRIORITY.
+
+#### ABNF
+```
+        "NO-NSW-xxx|xxx": {
+            "PACKET_ACTION": "FORWARD",
+            "IP_TYPE": "IPv6ANY",
+            "PRIORITY": "9xxxx20",
+            .....
+            ....
+        },
+```
+
+####  YANG
+```
+            leaf PACKET_ACTION {
+                type xxxx;
+            }
+            leaf IP_TYPE {
+                type xxxx;
+            }
+            leaf PRIORITY {
+                type uint32 {
+                    range xxxx;
+                }
+            }
+```
+
+
+### 3. Data Node/Object Hierarchy of the an objects in YANG models will be same as for all the objects at same hierarchy in Redis. If any exception is created then it must be recorded properly.
+For Example:
+
+"Family" of VLAN_INTERFACE and "IP_TYPE" of ACL_RULE should be at same level in YANG model too.
+
+#### ABNF
+```
+"VLAN_INTERFACE": {
+        "Vlan100|2a04:f547:45:6709::1/64": {
+            "scope": "global",
+            "family": "IPv6"
+        }
+}
+"ACL_RULE": {
+        "NO-NSW-PACL-V4|DEFAULT_DENY": {
+            "PACKET_ACTION": "DROP",
+            "IP_TYPE": "IPv4ANY",
+            "PRIORITY": "0"
+        }
+}
+```
+####  YANG
+In YANG, "Family" of VLAN_INTERFACE and "IP_TYPE" of ACL_RULE is at same level.
+```
+container VLAN_INTERFACE {
+        description "VLAN_INTERFACE part of config_db.json";
+        list xxx {
```

Comment:
Here we need to have a convention for list name. If I recall, from last meeting, we were in favor of something like <container name>_LIST, which in this case VLAN_INTERFACE_LIST

Reply:
We agreed to have a top level container also e.g. sonic-vlan. I will add in next PR on top of this.